### DISCUSSION TITLE
When V3?

### DISCUSSION THREAD

**Discussion by calebporzio** [livewire/livewire#5585] [8. V3 (maintainers only to post)]
Hi Livewire peeps.

**"When V3?"**
Right now I can't give you a date, but I CAN be transparent about the process and my progress, so here's what I would type to everyone who asks if I had time:

**"Why can't you set a date?"**
Because I don't know that I can stick to one and I don't want to say something and then go back on it. I COULD set a deadline for myself and ship no matter what, but I really don't want to be forced to skimp on things to get it out the door.

**"Deadlines and concessions are good when shipping software. Shipped is better than perfect"**
In general, I agree. If I ran a software company, I would want every developer to think like this. However, Livewire V3 is a ground-up rewrite that will likely be the codebase we will all work with and use for a very long time. If it's rushed and I don't like it, I won't be happy maintaining it and I won't be excited to add new features.

**"Aren't rewrites almost always a bad idea?"**
Yeah, almost always. And maybe even in this case too. In an alternate universe, I incrementally improved the codebase and shipped things as I went. However, Livewire is such a complex tool I'm not sure I would have been able to incrementally find better abstractions and performance enhancements without starting from scratch.

I'm trying to lay a foundation for a VERY bright future for Livewire. I want to take this tool as far as I possibly can, and this is the foundation of that.

**"Is Livewire really THAT complex of a tool?"**
I often find myself asking the same question. Hypothetically, one could write a super simple version of Livewire that would be extremely simple to build (I actually did this as a thought experiment in a Laracasts series called "Building Livewire from scratch"). But you would be missing key features that make Livewire worth using. For example: nesting components, `wire:model` binding, morphdom, security stuff, non-primitive public properties, etc... This list feels endless actually.

So again, yeah, I've asked myself this a thousand times and always come back with: yeah, it is. lol.

**"Why not involve the community more in this process?"**
I'm trying to do that more with this "V3" discussion section. Getting feedback from the community is super valuable to me, but opening up the codebase and accepting pull requests and such would actually make this process much slower. For now, I still want to be able to "move fast and break things" with the V3 codebase until I land on the right architecture and abstractions. So I will try to be better about looping in the community, but less on a: "here's the codebase, let's write V3 together" and more a "here's what I'm thinking about doing with X, what do you think?"

**"How much do you have left to do? I mean, you demoed a working version at Laracon right? Can't you just ship it?"**
I don't regret demoing it at Laracon, but I can see how it might feel weird that it's taking me so long.

What I demoed at Laracon was very experimental. It was a codebase that had all the V3 features, but only had hacky versions of the V2 features I needed for the talk.

The bulk of the work since has been supporting every single feature in V2 so that every single unit and browser test passes with the new codebase. We are nearly finished with that process, so now what's left is:

* Finalizing the new architecture. I'm mostly happy with it, but there are still things I'd like written more cleanly (particularly on the JavaScript end)
* Making the new V3 features production-ready. I had demo-ready versions built, but there's still some work to get them tested and fleshed out (particularly hot-reloading and "navigate")
* Documenting everything. I'm going to try and not do a complete rewrite of the docs because that would take forever, but I at least have to write new docs for all the new stuff and I want to re"work" all of the documentation to improve upon it in the ways you've all mentioned in the other post.
* Testing in actual applications
* Writing an upgrade guide (there are more changes than I anticipated, though still not an overwhelming amount)

**"When will you release a beta version?"**
I still don't have a date or anything, but I can say I've already put together an "alpha" and sent it to a few folks to try out. Once I have a "feature-complete" version for both V2 and V3 features I'm going to put it in "beta" and start sending it to more folks. Please hold off on asking for beta access until I'm at that point.

**"How will you launch it when it's time?"**
Well with lots of theatre of course! I'd like to do a small, online, event for the launch to properly show you everything live. That's just an idea right now and nothing more though.

---

So there you go, a giant thought spew of everything I think when someone says "when V3?".

TLDR: Can't give a date, but I can tell you I'm working every single day on getting it out the door.

**Comment by maxxmitchy** [comment]
Thanks Caleb. We appreciate the good work

**Comment by denizozturk** [comment]
The introduction of v3 was quite exciting, there are some exciting innovations worth waiting for. I hope you will complete it as soon as possible

**Comment by qwertynik** [comment]
Good things take time.
Would be interesting to read the vision you have for Livewire. For instance, is it going to be possible for someone to build a PWA app with it?

**Comment by utsavsomaiya** [comment]


**Reply by calebporzio** [reply]
Haha, wish I was there to see it.

**Comment by divdax** [comment]
We're all excited about what's to come, but we trust that you're taking the necessary time to do it right.

**Comment by iambateman** [comment]
You’re doing great, man! Grateful for you!

**Comment by Boorinio** [comment]
I get the no specific date ,but is there an approximation at least? a quarter or smth? :P

**Reply by edvordo** [reply]
I would guess same or very similar rules apply. Say he'd state V3 is gonna be done by the end of Q2 (end of june), but then mid june stumbles upon a thing that would break half the internet. Now the end Q2 approximate date isn't going to be fulfilled, might be moved to december (intentionally exaggerating here) and folks get angry (yes, angry, I've seen it happen several times, in software, games and in many more areas).

I'm excited for V3, but having no date, even an approximate one, is better in the long run.

**Comment by AyoobMH** [comment]
You are doing amazing, keep up the awesome work 🔥🔥🔥

**Comment by pasindujr** [comment]
> I'm trying to lay a foundation for a VERY bright future for Livewire. I want to take this tool as far as I possibly can, and this is the foundation of that.

Pretty excited about this. Take your time Caleb!

**Comment by michael-rubel** [comment]
@calebporzio Impressive that you're working on quality for the future! Don't hurry much. Such work should be done carefully.

I hope you also take a look into the issue of wire:model's rollbacks, as it is the major drawback of Livewire compared to the popular front-end frameworks. @joshhanley definitely knows what I'm talking about.

Once V3 is released, I'm eager to update the [Livewire Best Practices](https://github.com/michael-rubel/livewire-best-practices) 😉

**Comment by aurawindsurfing** [comment]
Every person needs to sponsor @calebporzio to let him make the best possible v3 and wait patiently.

**Comment by hilmanski** [comment]
take your time. Thanks for being transparent about this

**Comment by sarfraznawaz2005** [comment]
Your rational is correct as far as I am concerned. Thanks for such great package.

**Comment by brandon-9** [comment]
Keep it up. No rush.

**Comment by nickciolpan** [comment]
Your genuine and transparent approach is better than a deadline. Thank you for everything.

**Comment by mugerwa96** [comment]
Thnkx @calebporzio ,take your time, all what I know, livewire v3 will be better than the complex frameworks like vue and react looking to be easy hence maintaining two code bases which is difficult. With livewire, I maintain one code base, I will always pray for you  CALEB , you are the reason as to why I love laravel and the reason for exploring more things and never give up, thnkx alott

**Comment by KTanAug21** [comment]
"the codebase we will all work with and use for a very long time" -- that sounds _legendary_! The implementations I can do with v2 is already amazing in itself---v3 will come in the right time, there's no need to rush great work.

Thank you for creating Livewire @calebporzio, it's amazing! 🌟

**Comment by turbo124** [comment]
@calebporzio 

Love your work, please don't take this the wrong way, but perfection is a journey not a destination. When it feels right you'll know.

**Comment by BrendanTWhite** [comment]
For the record, I'm 190% team "no poll", and will be till the day I die.  ;-)

Also - something you love, even if it's a little later, is better than something you will always wish was different.

**Comment by jasonhuhx** [comment]
Respect!

**Comment by danve** [comment]
@calebporzio thank you,

**Comment by akr4m** [comment]
Thanks, @calebporzio for letting this know. I was desperately waiting for this.
I know L3 has huge complexity to develop. I respect what you are doing.

**Comment by claymore07** [comment]
We are grateful for your time and efforts, you have made great impact on the Laravel community.

**Comment by ChrisSantiago82** [comment]
People are asking for a date because they are excited about it. I think this is a good thing!
Of course, we want something good and we don't want to pressure you.
We are like kids at Christmas, waiting to open gifts...

**Comment by coolsam726** [comment]
Totally worth the wait. @calebporzio we are always grateful to you for working so hard to ease our lives. 🙌🏾

**Comment by Wulfheart** [comment]
I think the thing that most people got excited is that it was introduced half a year ago. 😂 
I can't wait to get my hands on it. Maybe you can share updates about the progress more regularly?

**Reply by mayaramyadav** [reply]
u are right bro..

**Comment by divoto** [comment]
Hi. I am sponsoring this project since the day I found how much fit to all I develop and complete replaced need to use Vue. As we don't hear many updates can we at least get a confirmation that there will be version 3 and work is continuing.

**Reply by Boorinio** [reply]
I think this discussion is already   Enough of an answer for what you are asking and the fact that he said there is a closed alpha circulating around. Although I agree a progress update would be really appreciated!!

**Reply by joshhanley** [reply]
@divoto thanks for your support 🙂 yep, we’re actively working on V3. It’s Caleb’s and my primary focus until it is released.

Can’t give you an update though, as that’s up to Caleb to decide, but rest assured that it’s being worked on 😁Best thing to do is follow Caleb on Twitter, as you’ll see he occasionally tweets about V3 related stuff as he is working on it.

**Reply by divoto** [reply]
Thanks for the good work and the update. Looking forward soon to get hands on V3.
Best regards

**Reply by Wulfheart** [reply]
I guess the community would be happier with more frequent updates.

**Comment by ariaieboy** [comment]
https://v3when.com/

**Reply by RealMrHex** [reply]
429 error

**Reply by joshhanley** [reply]
@RealMrHex check again 🙂

**Reply by RealMrHex** [reply]
@joshhanley Same error :(

**Reply by rustynail1984** [reply]
@RealMrHex you didnt get it? use your Brain dude xD

**Reply by RealMrHex** [reply]
> @RealMrHex you didnt get it? use your Brain dude xD

😂😂😂

**Comment by ariaieboy** [comment]
Only 45 days left until LaraconUS! It's possible that we'll receive some updates from the event. 🤞

**Reply by Wulfheart** [reply]
This would be awesome.

**Comment by joshtorres** [comment]
Thanks for the transparency and the hard work Caleb! My biggest question and concern, admittedly, is around sustainability. Do you have a team helping you? What can we do to help it be something sustainable for you to create and support it for years to come? Is simply supporting you on Github enough to ensure this is a product we can count on to be around for a while? As always, we appreciate everything you guys are doing. Livewire is one of the coolest things to come in the industry for a while, and I really hope it can find a solid foundation where large businesses can trust it's worth the investment for the next several years at least.

**Reply by ramonmalcolm10** [reply]
Absolutely, I share your concerns regarding both Livewire and Inertia Js. It would be greatly beneficial if these packages could adopt a model akin to Sveltekit, encouraging more community contributions to expedite the creation of innovative features. Moreover, the continuity and sustained evolution of these frameworks should be given paramount importance. Irrespective of future uncertainties in the tech landscape, it's vital that these frameworks continue to advance and maintain their relevance, providing a stable and reliable toolset for developers. Additionally, I would appreciate it if these packages could share their roadmaps and future plans. PS: Considering these points, I firmly believe that these packages harbor immense potential.

**Comment by ariaieboy** [comment]
Finally we have a public announcement 😍

>"When V3?"
...a question I've been asked a zillion times since giving a [sneak peek at Laracon](https://click.convertkit-mail.com/v8unnx7dqqurhxoper9hg/48hvhehrp3627lsx/aHR0cHM6Ly93d3cueW91dHViZS5jb20vbGl2ZS9mNFFTaEY0MmM2RT9mZWF0dXJlPXNoYXJlJnQ9NjAxNQ==) last year.
So many times, in fact, that someone made a website for it: [https://v3when.com](https://click.convertkit-mail.com/v8unnx7dqqurhxoper9hg/reh8hoh03pw5orc2/aHR0cHM6Ly92M3doZW4uY29tLw==)
Welp, I finally have an answer for you:
Livewire V3 will be publicly available after my Laracon talk on July 20th.
Version 3 of Livewire isn't a small improvement on V2; it's a complete rewrite. The code, the documentation, the website, everything.
It's taken me well over a year, and is still not quite finished. However, it's time that I commit to a deadline.
I'm so unbelievably stoked to finally get this out there and in your hands.
Cheers to a new era of Livewire,
-Caleb

**Reply by RealMrHex** [reply]
cant wait 😃😃😃

**Reply by aurawindsurfing** [reply]
🎉🎉🎉

**Comment by iammuttaqi** [comment]
Ah! Finally it's coming on July 20th..

**Comment by ramonmalcolm10** [comment]
Exciting times a head

**Comment by sbc640964** [comment]
Which devtools @calebporzio installed there in laracon US?

**Reply by nuernbergerA** [reply]
It's an yet internal package that will be released in the near future.

**Reply by sbc640964** [reply]
Not released yet? Or did I miss it?

**Reply by calebporzio** [reply]
Not released yet

**Comment by lokutus24** [comment]
hello,

i tried to install V3 according to documentation, but it is not working.
i click on the counter and it do nothing. it send not post request to the backend.
